client-sdk/ts-web/rt/src/callformat.ts

```diff
@@ -1,4 +1,92 @@
+import * as oasis from '@oasisprotocol/client';
+import * as ed from '@noble/ed25519';
+
+import * as transaction from './transaction';
+import * as mrae from './mrae';
+import * as types from './types';
+
 /**
  * Call data key pair ID domain separation context base.
  */
 export const CALL_DATA_KEY_PAIR_ID_CONTEXT_BASE = 'oasis-runtime-sdk/private: tx';
+
+/**
+ *  EncodeConfig is call encoding configuration.
+ *  golang: oasis-sdk/client-sdk/go/callformat/callformat.go
+ *  rust:
+ */
+export interface EncodeConfig {
+    /**
+     * publicKey is an optional runtime's call data public key to use for encrypted call formats.
+     */
+    publicKey?: types.KeyManagerSignedPublicKey;
+}
+
+export interface MetaEncryptedX25519DeoxysII {
+    sk: Uint8Array;
+    pk: Uint8Array;
+}
+
+/**
+ * encodeCall encodes a call based on its configured call format.
+ * It returns the encoded call and any metadata needed to successfully decode the result.
+ */
+export async function encodeCall(
+    nonce: Uint8Array,
+    sk: Uint8Array,
+    pk: Uint8Array,
+    call: types.Call,
+    format: types.CallFormat,
+    config?: EncodeConfig,
+): Promise<[types.Call, MetaEncryptedX25519DeoxysII?]> {
```

Comment:
Strictly speaking, you could use something like [this](https://www.typescriptlang.org/play?#code/FAOwhgtgpgzgDmAxlABAWSgFzAEzNlAb2BVJSgA84B7AJ0xUwE85UAFAGzAEsQUBeFCACuHDgG4SZSjXqMWqABoAmAKyqAjAE4AIlGoUmMAJLGBRKWTIBrKEwBcQ4RABGUWpKsBfSV+Ay6BmZWdCxcfDBzDGw8bAA6Th4+AB9QmIi4lXVtPQMjU0lQSFgEZBQAYTAxADE6CHwLKwC5YPYuXiiw2LAE9pBLcipA+RCszV19QxMzQWIrK2h07EdCIWLHAHIKNXGAWhxJo25uDZQfAasQahBkRwBVXkwADgBBWlowJk950lZ3e8er3en2+pD8TSGLQUKAA6gALfBQABu7nMcx+KEW4WWKA23Bw1g2oPm+OsjhgmFovAA5sS-H5mkFoZUanUGoIWRxarR6phMjscodpihUpzuby4vDESjaMBgAAzYQ3TDca7kG7UA4AHgAKoNMFAQDgYEQsd1HNFsWAvAA+AAUmEcOoAlE6ANoAIjNEQ9AF1GihaFhhLQ+HzvdhgH5-BqDnb0aQI2AVihbA4UAAWM4AGgGVxuUEcICgAHcUA8QM83h8mHa3RpsyhlI2AMy+525qx-WhF0vlwHVz51jON1SNgBs7dzXmdQA) to implement generic associated types. Go doesn't have that ability, which is why it's `interface{}`.

With the aim of not suggesting sweeping changes to the codebase, `unknown` is fine since at least it won't be a breaking change when other metadatas are added. It'll just be inconvenient to inspect the metadata.

Reply:
![image](https://user-images.githubusercontent.com/38229038/175995616-59b4df60-744c-4718-a3be-32e8fb6eed60.png)

I guess I'll check on that sample later today :speak_no_evil:

Reply:
ugh we used to do this thing in Go where there would be an overarching struct with a named field for each variant.

ok I loaded up the sample, but it's not clear how to go between associated Metadata.* and CallFormat.* type. (seems incidental and slightly abusive that CallFormat.Plain = Metadata.Plain.)

another answer would be to have multiple typescript declaration overloads of such an encode function. as nhynes says though, feel free not to and use `unknown` instead

I think it's fine that these types come out of encode relatively opaque. users will only pass them verbatim to submit and decode, respectively, neither of which internally could take advantage more type info